openapi-codegen/core/src/main/scala/sttp/tapir/codegen/ClassDefinitionGenerator.scala

```diff
@@ -3,96 +3,190 @@ package sttp.tapir.codegen
         case (name, obj: OpenapiSchemaEnum) =>
-          generateEnum(name, obj, targetScala3, queryParamRefs)
-        case (name, OpenapiSchemaMap(valueSchema, _)) => generateMap(name, valueSchema)
+          generateEnum(name, obj, targetScala3, queryParamRefs, jsonSerdeLib, allTransitiveJsonParamRefs)
+        case (name, OpenapiSchemaMap(valueSchema, _)) => generateMap(name, valueSchema, jsonSerdeLib, allTransitiveJsonParamRefs)
         case (n, x) => throw new NotImplementedError(s"Only objects, enums and maps supported! (for $n found ${x})")
       })
       .map(_.mkString("\n"))
-    defns.map(enumQuerySerdeHelper + _)
+    defns.map(additionalExplicitSerdes + maybeJsonSerdeHelpers + enumQuerySerdeHelper + _)
   }
 
-  private[codegen] def generateMap(name: String, valueSchema: OpenapiSchemaType): Seq[String] = {
+  private def enumQuerySerdeHelperDefn(targetScala3: Boolean): String = if (targetScala3)
+    """
+      |def enumMap[E: enumextensions.EnumMirror]: Map[String, E] =
+      |  Map.from(
+      |    for e <- enumextensions.EnumMirror[E].values yield e.name.toUpperCase -> e
+      |  )
+      |
+      |def makeQueryCodecForEnum[T: enumextensions.EnumMirror]: sttp.tapir.Codec[List[String], T, sttp.tapir.CodecFormat.TextPlain] =
+      |  sttp.tapir.Codec
+      |    .listHead[String, String, sttp.tapir.CodecFormat.TextPlain]
+      |    .mapDecode(s =>
+      |      // Case-insensitive mapping
+      |      scala.util
+      |        .Try(enumMap[T](using enumextensions.EnumMirror[T])(s.toUpperCase))
+      |        .fold(
+      |          _ =>
+      |            sttp.tapir.DecodeResult.Error(
+      |              s,
+      |              new NoSuchElementException(
+      |                s"Could not find value $s for enum ${enumextensions.EnumMirror[T].mirroredName}, available values: ${enumextensions.EnumMirror[T].values.mkString(", ")}"
+      |              )
+      |            ),
+      |          sttp.tapir.DecodeResult.Value(_)
+      |        )
+      |    )(_.name)
+      |""".stripMargin
+  else
+    """def makeQueryCodecForEnum[T <: enumeratum.EnumEntry](enumName: String, T: enumeratum.Enum[T]): sttp.tapir.Codec[List[String], T, sttp.tapir.CodecFormat.TextPlain] =
+      |  sttp.tapir.Codec.listHead[String, String, sttp.tapir.CodecFormat.TextPlain]
+      |    .mapDecode(s =>
+      |      // Case-insensitive mapping
+      |      scala.util.Try(T.upperCaseNameValuesToMap(s.toUpperCase))
+      |        .fold(
+      |          _ =>
+      |            sttp.tapir.DecodeResult.Error(
+      |              s,
+      |              new NoSuchElementException(
+      |                s"Could not find value $s for enum ${enumName}, available values: ${T.values.mkString(", ")}"
+      |              )
+      |            ),
+      |          sttp.tapir.DecodeResult.Value(_)
+      |        )
+      |    )(_.entryName)
+      |""".stripMargin
+
+  @tailrec
+  final def recursiveFindAllReferencedSchemaTypes(
```

Comment:
Part of the reason for farting about with this (admittedly pretty sweet) walk is to enable us to track usage sites for schema models, thus letting us choose the correct implementation for binary formats. Although not generating unused serdes is also nice.